airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/fileformatparser.py

```diff
@@ -132,49 +139,116 @@ def is_binary(self):
 
     def get_inferred_schema(self, file: Union[TextIO, BinaryIO]) -> dict:
         """
         https://arrow.apache.org/docs/python/generated/pyarrow.csv.open_csv.html
-        Note: this reads just the first block (as defined in _read_options() block_size) to infer the schema
+
+        This now uses multiprocessing in order to timeout the schema inference as it can hang.
+        Since the hanging code is resistant to signal interrupts, threading/futures doesn't help so needed to multiprocess.
+        https://issues.apache.org/jira/browse/ARROW-11853?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel
         """
-        streaming_reader = pa_csv.open_csv(file, self._read_options(), self._parse_options(), self._convert_options())
-        schema_dict = {field.name: field.type for field in streaming_reader.schema}
+
+        def get_schema(file_sample: str, read_opts: dict, parse_opts: dict, convert_opts: dict) -> dict:
+            """
+            we need to reimport here to be functional on Windows systems since it doesn't have fork()
+            https://docs.python.org/3.7/library/multiprocessing.html#contexts-and-start-methods
+            """
+            import tempfile
+
+            import pyarrow as pa
+
+            # writing our file_sample to a temporary file to then read in and schema infer as before
+            with tempfile.TemporaryFile() as fp:
+                fp.write(file_sample)
+                fp.seek(0)
+                streaming_reader = pa.csv.open_csv(
+                    fp, pa.csv.ReadOptions(**read_opts), pa.csv.ParseOptions(**parse_opts), pa.csv.ConvertOptions(**convert_opts)
+                )
+                schema_dict = {field.name: field.type for field in streaming_reader.schema}
+            return schema_dict
+
+        # boto3 stuff can't be pickled and so we can't multiprocess with the actual fileobject on Windows systems
+        # we're reading block_size*2 bytes here, which we can then pass in and infer schema from block_size bytes
+        # the *2 is to give us a buffer as pyarrow figures out where lines actually end so it gets schema correct
+        file_sample = file.read(self._read_options()["block_size"] * 2)
+        schema_dict = None
+        fail_count = 0
+        while schema_dict is None:
+            q_worker = mp.Queue()
+            proc = mp.Process(
+                target=multiprocess_queuer,
+                args=(
+                    dill.dumps(get_schema),  # use dill to pickle the get_schema function for Windows-compatibility
+                    q_worker,
+                    file_sample,
+                    self._read_options(),
+                    self._parse_options(),
+                    self._convert_options(),
+                ),
+            )
+            proc.start()
+            try:
+                # this attempts to get return value from function with a 20 second timeout
+                schema_dict = q_worker.get(timeout=20)
```

Comment:
it might be useful to do an increasing backoff until 60 seconds or something just in case the schema is genuinely taking a long time to load

Reply:
yeah I think the balancing act here is waiting long enough that we don't time out any long but non-hanging processes while failing relatively quickly in cases of hang or badly formed csvs that should fail inference.

Changed this to a doubling timeout from 4 -> 60 seconds

Reply:
Now propagating any actual non-hang errors back up now so it will failfast unless hanging in which case increasing backoff retries happen.